### PR TITLE
Graphite Datasource: add responseType: 'text' to http options to return full list of functions

### PR DESCRIPTION
**Pull request by bohandley**
This fix is allows the graphite datasource to successfully return a full list of functions. 

#### What is the problem? 
https://github.com/grafana/observability-metrics-squad/issues/16

In a previous update to the core fetch functionality, https://github.com/grafana/grafana/pull/45188, we began passing in responseType which we had not been passing in before. The graphite response type was previously undefined which then defaulted to json based on the change by @leeoniya. 

Returning the graphite response.json() results in an error due to graphite returning large numbers as the keyword Infinity, https://raintank-corp.slack.com/archives/C0368BHNQ79. A week or so after Leon's fix, there was this change to fetch, https://github.com/grafana/grafana/pull/45598, which caught the error and returned an empty object. This resulted in no functions being returned and was fixed by this change https://github.com/grafana/grafana/commit/ed155e47c4f9a2c57acfee0a9a5ca681b5dd2894 by @ifrost.

There is currently another change by @leventebalogh, https://github.com/grafana/grafana/pull/47493. This change removes the error handling for json response types in the fetch parse response body function. This then causes the graphite datasource to receive the error from response.json() and eventually returns a default list of functions due to this https://github.com/grafana/grafana/blob/a4381ebc9122d121e80cb7636c69c408422271fe/public/app/plugins/datasource/graphite/datasource.ts#L783-L787. @ryantxu made the suggestion we set the responseType to text to handle this.

##### The fix
So, this fix sets the responseType to text so that we bypass the json issue in fetch parseResponseBody and then allows the workaround for the graphite Infinity issue here https://github.com/grafana/grafana/blob/a4381ebc9122d121e80cb7636c69c408422271fe/public/app/plugins/datasource/graphite/datasource.ts#L757-L765.

### CODE REVIEW

File: public/app/plugins/datasource/graphite/datasource_integration.test.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
+import { GraphiteDatasource } from './datasource';
+import { TemplateSrv } from 'app/features/templating/template_srv';
+import { BackendSrv } from 'app/core/services/backend_srv';
+import { of } from 'rxjs';
+import { ContextSrv, User } from '../../../core/services/context_srv';
+import { setBackendSrv } from '@grafana/runtime';
+
+jest.mock('@grafana/runtime', () => ({
+  ...(jest.requireActual('@grafana/runtime') as unknown as object),
```
Comment (aocenas) [inline public/app/plugins/datasource/graphite/datasource_integration.test.ts:9]: Wonder why is this here? If I understand this does sort of nothing as it mocks runtime with itself?

Comment (bohandley) [reply public/app/plugins/datasource/graphite/datasource_integration.test.ts:9]: I see, yes, we don't need this. I saw that it had been used in /app/plugins/datasource/graphite/datasource.test.ts but I think this was to replace the backendSrv with the imported version in @grafana/runtime. In the integration test, it doesn't do anything so I'll take it out.

File: public/app/plugins/datasource/graphite/datasource.ts [hunk 2]
```diff
@@ -749,6 +749,8 @@ export class GraphiteDatasource
     const httpOptions = {
       method: 'GET',
       url: '/functions',
+      // to work around Graphite returning invalid JSON
+      responseType: 'text',
```
[the remark this answers is not in the document]
Comment (bohandley) [reply public/app/plugins/datasource/graphite/datasource.ts:753]: I believe is true, we should never get ```typeof results.data === 'object' && results.status === 200``` with responseType text. We will also never get the empty object, and that was being handled on line and 780. It looks like we will always get a string too, as the response is text... so I think it seems safe to remove all the conditionals and just do the fix for Infinity, parse the functions and return them like this:

```
const fixedData = JSON.parse(results.data.replace(/"default": ?Infinity/g, '"default": 1e9999'));
this.funcDefs = gfunc.parseFuncDefs(fixedData);
return this.funcDefs;
```

[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/plugins/datasource/graphite/datasource.ts:753]: Yes, you can remove the fix for an empty object along with the test. In other words, all changes from ed155e47c4f9a2c57acfee0a9a5ca681b5dd2894 can be reverted.

File: public/app/plugins/datasource/graphite/datasource_integration.test.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    const instanceSettings = {
+      url: '/api/datasources/proxy/1',
+      name: 'graphiteProd',
+      jsonData: {
+        rollupIndicatorEnabled: true,
+      },
+    };
+    const templateSrv = new TemplateSrv();
+    const ds = new GraphiteDatasource(instanceSettings, templateSrv);
+    ctx = { templateSrv, ds };
```
Comment (ifrost) [inline public/app/plugins/datasource/graphite/datasource_integration.test.ts:32]: `TemplateSrv` is created as a default value in the constructor, and it doesn't seem to be used in the test so you can also simplify it to:

```suggestion
    const ds = new GraphiteDatasource(instanceSettings);
    ctx = { ds };
```

File: public/app/plugins/datasource/graphite/datasource_integration.test.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+              type: 'int_or_infinity',
+            },
+          ],
+        },
+      });
+    });
+  });
+});
+
+function mockBackendSrv(data: string) {
```
Comment (ifrost) [inline public/app/plugins/datasource/graphite/datasource_integration.test.ts:94]: I think this might be too excessive. It should be enough to mock the fetch with the following:

```typescript
function mockBackendSrv(data: string) {
  const fetchMock = jest.fn().mockReturnValue(of({ data }));
  setBackendSrv({ fetch: fetchMock } as any);
}
```

**Review by ifrost — CHANGES_REQUESTED** [review]

File: public/app/plugins/datasource/graphite/datasource_integration.test.ts (hunk 3 above)
Comment (bohandley) [reply public/app/plugins/datasource/graphite/datasource_integration.test.ts:32]: Can do!

File: public/app/plugins/datasource/graphite/datasource_integration.test.ts (hunk 4 above)
Comment (bohandley) [reply public/app/plugins/datasource/graphite/datasource_integration.test.ts:94]: I hear you on it being excessive, but we wanted to test the fromFetch dependency in backendSrv, which would then call parseResponseBody from /core/utils/fetch.ts. By only mocking the fetch, we don't get the full integration test with parseResponseBody.

Comment (ifrost) [reply public/app/plugins/datasource/graphite/datasource_integration.test.ts:89]: Oh yes, of course 🤦  Why do we need to mock `parseRequestOptionsMock` though?

File: public/app/plugins/datasource/graphite/datasource.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (aocenas) [reply public/app/plugins/datasource/graphite/datasource.ts:754]: > Are we allowed to remove tests?

Yes definitely, if they don't make sense anymore.

[the remark this answers is not in the document]
Comment (bohandley) [reply public/app/plugins/datasource/graphite/datasource.ts:754]: Thank you! I removed the test for the empty object.

File: public/app/plugins/datasource/graphite/datasource_integration.test.ts (hunk 4 above)
Comment (bohandley) [reply public/app/plugins/datasource/graphite/datasource_integration.test.ts:89]: That's true! we don't need ```parseRequestOptions```, since the options are already in the mocked fromFetch. I removed the mock for ```parseRequestOptions```.

**Review by ifrost — APPROVED** [review]

File: public/app/plugins/datasource/graphite/datasource.ts [hunk 5]
```diff
@@ -750,35 +751,20 @@ export class GraphiteDatasource
...
-          } else {
-            this.funcDefs = gfunc.parseFuncDefs(results.data);
-          }
-
-          // When /functions endpoint returns application/json response but containing invalid JSON the fix above
-          // wont' be triggered due to the changes in https://github.com/grafana/grafana/pull/45598 (parsing happens
-          // in fetch and Graphite receives an empty object and no error). In such cases, when the provided JSON
-          // seems empty we fallback to the hardcoded list of functions.
-          // See also: https://github.com/grafana/grafana/issues/45948
-          if (Object.keys(this.funcDefs).length === 0) {
```
Comment (aocenas) [inline public/app/plugins/datasource/graphite/datasource.ts:779]: @Dieterbe is there a case when graphite API itself would return empty list of functions and it would make sense for Grafana to supply some defaults?

This here is removed because because we circumvent the changed behaviour that invalidated the workaround and we always get the response as text and we can then apply workaround for the `infinity` issue. But that means if Graphite returns empty list of functions we assume there are no functions in Graphite. Is that valid or will Graphite always return at least the default function set?

Comment (Dieterbe) [reply public/app/plugins/datasource/graphite/datasource.ts:779]: Graphite should never return an empty list of functions, unless something went badly wrong with it :) 
Note: ownership of our graphite product has been moved to the query squad. i'm no longer involved.

**Review by aocenas — APPROVED** [review]
